File: buyer-ios/PoultryBuyer/Features/Search/ViewModels/SearchViewModel.swift
import Foundation
import Combine

@MainActor
class SearchViewModel: ObservableObject {
    @Published var query: String = ""
    @Published var products: [Product] = []
    @Published var categories: [Category] = []
    @Published var selectedCategoryId: String? = nil
    @Published var isLoading = false
    @Published var isLoadingMore = false
    @Published var error: String? = nil
    @Published var hasMore = true

    private var currentPage = 0
    private let pageSize = 20
    private var searchTask: Task<Void, Never>?

    init() {
        loadCategories()
    }

    func loadCategories() {
        Task {
            do {
                let response = try await APIClient.shared.getCategories()
                if response.success, let data = response.data {
                    categories = data
                }
            } catch {
                // Categories are optional, don't show error
            }
        }
    }

    func onQueryChange(_ newQuery: String) {
        query = newQuery

        // Debounce search
        searchTask?.cancel()
        searchTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000) // 300ms debounce

            if Task.isCancelled { return }

            if !query.isEmpty {
                await searchProducts(resetPage: true)
            } else if selectedCategoryId != nil {
                await loadProductsByCategory(resetPage: true)
            } else {
                products = []
            }
        }
    }

    func onCategorySelected(_ categoryId: String?) {
        selectedCategoryId = categoryId
        query = ""

        if let categoryId = categoryId {
            Task {
                await loadProductsByCategory(resetPage: true)
            }
        } else {
            products = []
        }
    }

    func loadMore() {
        guard !isLoading, !isLoadingMore, hasMore else { return }

        Task {
            if !query.isEmpty {
                await searchProducts(resetPage: false)
            } else if selectedCategoryId != nil {
                await loadProductsByCategory(resetPage: false)
            }
        }
    }

    func refresh() async {
        if !query.isEmpty {
            await searchProducts(resetPage: true)
        } else if selectedCategoryId != nil {
            await loadProductsByCategory(resetPage: true)
        }
    }

    private func searchProducts(resetPage: Bool) async {
        let page = resetPage ? 0 : currentPage + 1

        if resetPage {
            isLoading = true
        } else {
            isLoadingMore = true
        }
        error = nil

        do {
            let response = try await APIClient.shared.searchProducts(query: query, page: page, size: pageSize)
            if response.success, let data = response.data {
                if resetPage {
                    products = data.content
                } else {
                    products.append(contentsOf: data.content)
                }
                currentPage = page
                hasMore = !data.last
            } else {
                error = response.message ?? "Search failed"
            }
        } catch {
            self.error = "Network error. Please try again."
        }

        isLoading = false
        isLoadingMore = false
    }

    private func loadProductsByCategory(resetPage: Bool) async {
        guard let categoryId = selectedCategoryId else { return }

        let page = resetPage ? 0 : currentPage + 1

        if resetPage {
            isLoading = true
        } else {
            isLoadingMore = true
        }
        error = nil

        do {
            let response = try await APIClient.shared.getProductsByCategory(categoryId: categoryId, page: page, size: pageSize)
            if response.success, let data = response.data {
                if resetPage {
                    products = data.content
                } else {
                    products.append(contentsOf: data.content)
                }
                currentPage = page
                hasMore = !data.last
            } else {
                error = response.message ?? "Failed to load products"
            }
        } catch {
            self.error = "Network error. Please try again."
        }

        isLoading = false
        isLoadingMore = false
    }

    func clearError() {
        error = nil
    }
}

File: buyer-android/app/src/main/java/com/poultry/buyer/ui/screens/search/SearchViewModel.kt
package com.poultry.buyer.ui.screens.search

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.poultry.buyer.core.network.ApiService
import com.poultry.buyer.domain.model.Category
import com.poultry.buyer.domain.model.Product
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SearchUiState(
    val query: String = "",
    val products: List<Product> = emptyList(),
    val categories: List<Category> = emptyList(),
    val selectedCategoryId: String? = null,
    val isLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val error: String? = null,
    val currentPage: Int = 0,
    val hasMore: Boolean = true
)

@HiltViewModel
class SearchViewModel @Inject constructor(
    private val apiService: ApiService
) : ViewModel() {

    private val _uiState = MutableStateFlow(SearchUiState())
    val uiState: StateFlow<SearchUiState> = _uiState.asStateFlow()

    private var searchJob: Job? = null

    init {
        loadCategories()
    }

    private fun loadCategories() {
        viewModelScope.launch {
            try {
                val response = apiService.getCategories()
                if (response.isSuccessful && response.body()?.success == true) {
                    _uiState.value = _uiState.value.copy(
                        categories = response.body()?.data ?: emptyList()
                    )
                }
            } catch (e: Exception) {
                // Categories are optional, don't show error
            }
        }
    }

    fun onQueryChange(query: String) {
        _uiState.value = _uiState.value.copy(query = query)

        // Debounce search
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            delay(300) // 300ms debounce
            if (query.isNotBlank()) {
                searchProducts(query, resetPage = true)
            } else if (_uiState.value.selectedCategoryId != null) {
                loadProductsByCategory(_uiState.value.selectedCategoryId!!, resetPage = true)
            } else {
                _uiState.value = _uiState.value.copy(products = emptyList())
            }
        }
    }

    fun onCategorySelected(categoryId: String?) {
        _uiState.value = _uiState.value.copy(
            selectedCategoryId = categoryId,
            query = ""
        )

        if (categoryId != null) {
            loadProductsByCategory(categoryId, resetPage = true)
        } else {
            _uiState.value = _uiState.value.copy(products = emptyList())
        }
    }

    fun loadMore() {
        if (_uiState.value.isLoading || _uiState.value.isLoadingMore || !_uiState.value.hasMore) return

        val state = _uiState.value
        if (state.query.isNotBlank()) {
            searchProducts(state.query, resetPage = false)
        } else if (state.selectedCategoryId != null) {
            loadProductsByCategory(state.selectedCategoryId, resetPage = false)
        }
    }

    fun refresh() {
        val state = _uiState.value
        if (state.query.isNotBlank()) {
            searchProducts(state.query, resetPage = true)
        } else if (state.selectedCategoryId != null) {
            loadProductsByCategory(state.selectedCategoryId, resetPage = true)
        }
    }

    private fun searchProducts(query: String, resetPage: Boolean) {
        viewModelScope.launch {
            val page = if (resetPage) 0 else _uiState.value.currentPage + 1

            _uiState.value = _uiState.value.copy(
                isLoading = resetPage,
                isLoadingMore = !resetPage,
                error = null
            )

            try {
                val response = apiService.searchProducts(query, page, PAGE_SIZE)
                if (response.isSuccessful && response.body()?.success == true) {
                    val data = response.body()?.data
                    val newProducts = data?.content ?: emptyList()

                    _uiState.value = _uiState.value.copy(
                        products = if (resetPage) newProducts else _uiState.value.products + newProducts,
                        currentPage = page,
                        hasMore = !(data?.last ?: true),
                        isLoading = false,
                        isLoadingMore = false
                    )
                } else {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        isLoadingMore = false,
                        error = response.body()?.message ?: "Search failed"
                    )
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    isLoadingMore = false,
                    error = "Network error. Please try again."
                )
            }
        }
    }

    private fun loadProductsByCategory(categoryId: String, resetPage: Boolean) {
        viewModelScope.launch {
            val page = if (resetPage) 0 else _uiState.value.currentPage + 1

            _uiState.value = _uiState.value.copy(
                isLoading = resetPage,
                isLoadingMore = !resetPage,
                error = null
            )

            try {
                val response = apiService.getProductsByCategory(categoryId, page, PAGE_SIZE)
                if (response.isSuccessful && response.body()?.success == true) {
                    val data = response.body()?.data
                    val newProducts = data?.content ?: emptyList()

                    _uiState.value = _uiState.value.copy(
                        products = if (resetPage) newProducts else _uiState.value.products + newProducts,
                        currentPage = page,
                        hasMore = !(data?.last ?: true),
                        isLoading = false,
                        isLoadingMore = false
                    )
                } else {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        isLoadingMore = false,
                        error = response.body()?.message ?: "Failed to load products"
                    )
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    isLoadingMore = false,
                    error = "Network error. Please try again."
                )
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    companion object {
        private const val PAGE_SIZE = 20
    }
}
